gtk/src/file_filter_info.rs

```diff
@@ -0,0 +1,176 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use std::ffi::CStr;
+
+use glib::translate::*;
+use libc::c_char;
+
+use crate::FileFilterFlags;
+
+#[repr(transparent)]
+#[doc(alias = "GtkFileFilterInfo")]
+pub struct FileFilterInfo(ffi::GtkFileFilterInfo);
+
+impl FileFilterInfo {
+    pub fn new(
+        filename: Option<&str>,
+        uri: Option<&str>,
+        display_name: Option<&str>,
+        mime_type: Option<&str>,
+    ) -> Self {
+        skip_assert_initialized!();
+        let mut contains = FileFilterFlags::empty();
+        for (value, bit) in [
+            (filename, FileFilterFlags::FILENAME),
+            (uri, FileFilterFlags::URI),
+            (display_name, FileFilterFlags::DISPLAY_NAME),
+            (mime_type, FileFilterFlags::MIME_TYPE),
+        ] {
+            if value.is_some() {
+                contains |= bit;
+            }
+        }
+
+        Self(ffi::GtkFileFilterInfo {
+            contains: contains.bits(),
+            filename: filename.to_glib_full(),
+            uri: uri.to_glib_full(),
+            display_name: display_name.to_glib_full(),
+            mime_type: mime_type.to_glib_full(),
+        })
+    }
+
+    #[doc(alias = "get_contains")]
+    pub fn contains(&self) -> FileFilterFlags {
+        FileFilterFlags::from_bits_truncate(self.0.contains)
+    }
+
+    #[doc(alias = "get_filename")]
+    pub fn filename(&self) -> Option<&str> {
+        if self.0.filename.is_null() {
+            None
+        } else {
+            unsafe { CStr::from_ptr(self.0.filename).to_str().ok() }
+        }
+    }
+
+    #[doc(alias = "get_uri")]
+    pub fn uri(&self) -> Option<&str> {
+        if self.0.uri.is_null() {
+            None
+        } else {
+            unsafe { CStr::from_ptr(self.0.uri).to_str().ok() }
+        }
+    }
+
+    #[doc(alias = "get_display_name")]
+    pub fn display_name(&self) -> Option<&str> {
+        if self.0.display_name.is_null() {
+            None
+        } else {
+            unsafe { CStr::from_ptr(self.0.display_name).to_str().ok() }
+        }
+    }
+
+    #[doc(alias = "get_mime_type")]
+    pub fn mime_type(&self) -> Option<&str> {
+        if self.0.mime_type.is_null() {
+            None
+        } else {
+            unsafe { CStr::from_ptr(self.0.mime_type).to_str().ok() }
+        }
+    }
+}
+
+impl Drop for FileFilterInfo {
+    fn drop(&mut self) {
+        for ptr in [
+            self.0.filename,
+            self.0.uri,
+            self.0.display_name,
+            self.0.mime_type,
+        ] {
+            if !ptr.is_null() {
+                unsafe {
+                    glib::ffi::g_free(ptr as *mut _);
```

Comment:
`g_free()` can be called with `NULL` so that check is unneeded

Reply:
Is that safe?  It appears `g_free()` just calls the system `free()`.  Glibc does promise that passing `NULL` won't do anything, but do we believe that is true for all libcs (like musl, or non-Linux OSes?).

Reply:
C99 (at least) allows passing `NULL` to `free()`.